lib/Conversion/TorchToLinalg/AdaptivePooling.cpp

```diff
@@ -0,0 +1,237 @@
+#include "torch-mlir/Conversion/Utils/Utils.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchDialect.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/TorchUpstream.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+/*
+This file is for lowering adaptive pooling ops, which cannot generally be
+decomposed into typical pooling ops. Given an input tensor of rank (N,C,Hin) and
+an output spatial size Hout, an element of the output tensor at position (n, c,
+h) is computed as follows.
+    1. compute st(h) = (h*Hin)//Hout
+    2. compute en(h) = 1 + ((h+1)*Hin - 1)//Hout
+    3. apply the operation (max or avg) over input[n, c, st(h):en(h)]
+This is problematic for linalg ops for a few reasons:
+    1. The access to the input tensor is not constantly strided
+    2. The size of the window itself is not contant: en(h) - st(h) can vary with
+h! Although it is a bit like using a hammer to paint, our workaround is to use
+tensor.extract to access the elements of the input tensor inside our linalg
+generic op's payload.
+
+Current TODO's:
+    1. roll st(h) and en(h) into affine maps for aesthetics
+    2. gather most of the boilerplate out of this op and make it into an
+adaptive pooling helper function.
+    3. figure out what to do with the conflicting decompositions in
+DecomposeComplexOps.cpp
+    4. Implement more efficient passes for when the kernel-size, input spatial
+dims, and output spatial dims are constant.
+*/
+
+namespace {
+class ConvertAtenAdaptiveAvgPool1dOp
+    : public OpConversionPattern<AtenAdaptiveAvgPool1dOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenAdaptiveAvgPool1dOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Location loc = op->getLoc();
+    const TypeConverter *typeconverter = getTypeConverter();
+
+    // input tensor and output shape
+    Value input = adaptor.getSelf();
+    Value outputShape = adaptor.getOutputSize();
+    SmallVector<Value> outShapeVector;
+    getListConstructElements(outputShape, outShapeVector);
+    outShapeVector =
+        getTypeConvertedValues(rewriter, loc, typeconverter, outShapeVector);
+    Value Hin = getDimOp(rewriter, loc, input, 2);
+    Value Hout = outShapeVector[0];
+    Value HoutIndex = castIntToIndex(rewriter, loc, Hout);
+    RankedTensorType InputType = input.getType().cast<RankedTensorType>();
+    RankedTensorType OutputType =
+        typeconverter->convertType(op.getResult().getType())
+            .cast<RankedTensorType>();
+
+    // get rank of input (same as rank of output)
+    int64_t rank = input.getType().cast<RankedTensorType>().getRank();
+    // input operand should be NCH (i.e. rank 3)
+    if (rank != 3) {
+      return rewriter.notifyMatchFailure(op, "only supports input type NCH");
+    }
```

Comment:
When possible, error checking should happen _before_ any IR is created (functions that take a `rewriter` as an argument, like `getDimOp`, generate IR). Otherwise we risk leaving the graph in an invalid state after a match failure. This check should be one of the very first things this patterns does.